### DISCUSSION TITLE
OOM of Image automation controller

### DISCUSSION THREAD

**Discussion by ZeroDeltaAlpha** [fluxcd/flux2#2261] [Q&A] [answered]
### Problem
We are seeing the Ephemeral Storage of the image automation controller grow quite rapidly and, in the past have experienced OOM evictions due to this limit being reached. We have now given the controller a 100G limit as it routinely consumes 30G~ at any given time

What solutions are there other than feeding the deployment more ephemeral storage ?

### Specs and setup
```
flux: v0.24.1
helm-controller: v0.12.0
image-automation-controller: v0.15.0
image-reflector-controller: v0.12.0
kustomize-controller: v0.15.2
notification-controller: v0.17.0
source-controller: v0.16.0
```

7 Git repositories with 2 sources of them watching two branches for each repository.

**Comment by stefanprodan** [comment, marked as answer]
Is this the case with source-controller?

**Reply by ZeroDeltaAlpha** [reply]
Hi @stefanprodan thanks for getting back to me so fast, I suspect so. We get floods of 404's in slack when the tar.gz artifacts are gone due to the source controller being killed and this is when i think it is OM killed but it is a much slower process then the image-automation OOM

**Reply by stefanprodan** [reply]
Can you please exec in both controllers and run `ls -lah /tmp`

**Reply by squaremo** [reply]
(or `du -sh /tmp/*`, to get the sum of directory contents)

**Reply by ZeroDeltaAlpha** [reply]
So doing a du of tmp on image automation controller shows 41G and the source controller shows only 256M. I'm attaching a graph from grafana that show the gradual climb of the image controller.

![2022-01-06_09-18](https://user-images.githubusercontent.com/70947479/148360221-3be32d72-02a7-4781-b50b-c8152ce4b100.png)

**Reply by stefanprodan** [reply]
Looks like the controllers are not able to clean tmp after a clone operation, I don't see this behaviour on GKE/EKS nor Kubernetes Kind. Can you post here `kubectl version` and tell us more about how the node disks are managed, is the EmptyDir set to a RAM disk?

**Reply by ZeroDeltaAlpha** [reply]
Hi @stefanprodan The kubectl version is as follows
```
Client Version: v1.22.4
Server Version: v1.21.2-eks-0389ca3
```
I'm unsure of how the EmptyDir is being satisfied for backing storage as it its being created by the gotk-components.yaml in our flux repo. Strangely the graphs for source controller show up and down fluctuations like garbage collection is kicking in.

![image](https://user-images.githubusercontent.com/70947479/148363204-88fb5c24-e447-4ea7-9669-9294d7d4da3a.png)

**Reply by squaremo** [reply]
The source controller keeps the most recent tarballs around, so I would expect it to be using some disk space, always. The image-automation-controller always throws its working directory away, so should be empty most of the time. It certainly looks like it's failing to remove the working directory, and eating disk space as a result.

This PR https://github.com/fluxcd/image-automation-controller/pull/287 makes the controller log any error it gets when trying to remove the working directory (it was expected to always succeed :-/). If you wait until tomorrow, there'll be a nightly build with that change, which you could substitute in a dev environment, if you have one.

**Reply by ZeroDeltaAlpha** [reply]
Thanks @squaremo I'll give that a go when it's released and report back any findings.

**Reply by squaremo** [reply]
Stefan showed me how to do an RC release, so I've run that and published an image: `docker.io/fluxcd/image-automation-controller:rc-90dcdfd7` (as reported at https://github.com/fluxcd/image-automation-controller/runs/4725996812?check_suite_focus=true#step:9:4622). You can patch the image used for image-automation-controller in flux-system/kustomization.yaml, something like this:

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- gotk-components.yaml
- gotk-sync.yaml
images:
  - name: fluxcd/image-automation-controller
    newName: fluxcd/image-automation-controller
    newTag: rc-90dcdfd7
```

**Reply by ZeroDeltaAlpha** [reply]
So we have deployed the RC and the error we are getting in the logs is
```
unable to clone: SSH could not read data: Error waiting on socket
```
I Don't think this was a previous issue is there anything we are missing ?

**Reply by squaremo** [reply]
That's something people see from time to time (e.g., https://github.com/fluxcd/image-automation-controller/issues/286), but it's not a message I'd expect to be logged when the controller fails to rm from /tmp. Does that container have GB of files in /tmp?

**Reply by ZeroDeltaAlpha** [reply]
running du a few times it maxes out at 4G used then comes down to 2.7G and moves between those values so it does appear to be garbage collecting

**Reply by squaremo** [reply]
Are there any log messages containing the message `failed to remove working directory`?

**Reply by ZeroDeltaAlpha** [reply]
Nothing so far just those SSH errors, looking at the new copy it looks a lot less aggressive in filling up the disk.

**Reply by squaremo** [reply]
How's it look a few days later -- does it look unbounded but slower, or has the problem gone away?

**Reply by ZeroDeltaAlpha** [reply]
HI @squaremo , Seeing a slower rate of disk usage, Still getting Evicted but now due to DiskPressure on the node, would there be any down side to setting the tmp directory to use a EBS volume ? As I'm understanding that the controller does its own GC

**Reply by squaremo** [reply]
A couple of downsides:
 - putting git checkouts on persistent storage means they are more accessible through side channels (e.g., by mounting that EBS volume in some VM); if everyone with access to your cluster (and AWS infrastructure) is trusted, this may be less of a concern for you
 - If it grows without (apparent) bound in memory, it'll grow without bound in block storage -- though you will have a lot more headroom.

The "steady state" for the image-automation-controller is to be using no disk, so ultimately there's a bug there to fix. In the meantime, you could perhaps have fewer Evictions -- and thereby fewer automation outages and alerts -- if you use an EBS volume (modulo the concerns above).

**Reply by ZeroDeltaAlpha** [reply]
@squaremo to give you an idea of how our testing is going this copy of automation controller has been running for 3d4h and this is how it's ephemeral storage looks in grafana

![image](https://user-images.githubusercontent.com/70947479/149800112-c3b7a9f7-c2c5-481d-bb17-b28b4c63633c.png)

The GC looks more aggressive this time and doesn't breach 2G of usage, which is a massive improvement and we haven't had the pod get evicted off the node for any reason.

**Reply by kingdonb** [reply]
It sounds like things turned around. I just want to highlight for anyone that sees a recurrence of this issue,

* https://github.com/fluxcd/image-automation-controller/pull/287

added logging which should make it possible to pinpoint times when removing a working directory with `RemoveAll` failed.
